core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -255,17 +297,128 @@ private void rebuildIndex() throws Exception {
       }
    }
 
+   // Initialise missing internal metadata state for corrupt data
+   private PrivateMetadata generateMissingInternalMetadata(K key) {
+      // Optimistic Transactions
+      AdvancedCache<?, ?> cache = ctx.getCache().getAdvancedCache();
+      ComponentRegistry componentRegistry = cache.getComponentRegistry();
+      Configuration config = cache.getCacheConfiguration();
+      TransactionConfiguration txConfig = config.transaction();
+      PrivateMetadata.Builder builder = new PrivateMetadata.Builder();
+      if (txConfig.transactionMode() == TransactionMode.TRANSACTIONAL && txConfig.lockingMode() == LockingMode.OPTIMISTIC) {
+         builder.entryVersion(new NumericVersionGenerator().generateNew());
+      }
+
+      // Async XSite
+      if (config.sites().hasAsyncEnabledBackups()) {
+         KeyPartitioner keyPartitioner = componentRegistry.getComponent(KeyPartitioner.class);
+         IracVersionGenerator ivg = componentRegistry.getIracVersionGenerator().running();
+         int segment = keyPartitioner.getSegment(key);
+         builder.iracMetadata(ivg.generateMetadataWithCurrentVersion(segment));
```

Comment:
```suggestion
         String siteName = cache.getCacheManager().getTransport().localSiteName();
         IracEntryVersion version = new IracEntryVersion(Collections.singletonMap(siteName, TopologyIracVersion.newVersion(1)));
         IracMetadata metadata = new IracMetadata(siteName, version);
         builder.iracMetadata(metadata);
```

Reply:
Isn't this bypassing the stored global state?

Reply:
yes. unfortunately, I don't see a way of breaking the cyclic dependency 
We need to document this. The user must trigger the state transfer both ways; this will trigger a conflict and the merge policy should fix it and assign the same version on both sides for each key.